ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -33,119 +34,156 @@ import           Language.LSP.Server            (ProgressAmount (..),
+                                                 bracket, cancel)
 
 data ProgressEvent
-    = KickStarted
-    | KickCompleted
+  = ProgressStarted
+  | ProgressCompleted
 
-data ProgressReporting  = ProgressReporting
-  { progressUpdate :: ProgressEvent -> IO ()
-  , inProgress     :: forall a. NormalizedFilePath -> Action a -> Action a
-  , progressStop   :: IO ()
+data ProgressReporting m = ProgressReporting
+  { progressUpdate :: ProgressEvent -> IO (),
+    inProgress     :: forall a. NormalizedFilePath -> m a -> m a,
+    progressStop   :: IO ()
   }
 
-noProgressReporting :: IO ProgressReporting
-noProgressReporting = return $ ProgressReporting
-  { progressUpdate = const $ pure ()
-  , inProgress = const id
-  , progressStop   = pure ()
-  }
+noProgressReporting :: IO (ProgressReporting m)
+noProgressReporting =
+  return $
+    ProgressReporting
+      { progressUpdate = const $ pure (),
+        inProgress = const id,
+        progressStop = pure ()
+      }
 
 -- | State used in 'delayedProgressReporting'
 data State
-    = NotStarted
-    | Stopped
-    | Running (Async ())
+  = NotStarted
+  | Stopped
+  | Running (Async ())
 
 -- | State transitions used in 'delayedProgressReporting'
 data Transition = Event ProgressEvent | StopProgress
 
 updateState :: IO () -> Transition -> State -> IO State
-updateState _      _                    Stopped       = pure Stopped
-updateState start (Event KickStarted)   NotStarted    = Running <$> async start
-updateState start (Event KickStarted)   (Running job) = cancel job >> Running <$> async start
-updateState _     (Event KickCompleted) (Running job) = cancel job $> NotStarted
-updateState _     (Event KickCompleted) st            = pure st
-updateState _     StopProgress          (Running job) = cancel job $> Stopped
-updateState _     StopProgress          st            = pure st
+updateState _ _ Stopped = pure Stopped
+updateState start (Event ProgressStarted) NotStarted = Running <$> async start
+updateState start (Event ProgressStarted) (Running job) = cancel job >> Running <$> async start
+updateState _ (Event ProgressCompleted) (Running job) = cancel job $> NotStarted
+updateState _ (Event ProgressCompleted) st = pure st
+updateState _ StopProgress (Running job) = cancel job $> Stopped
+updateState _ StopProgress st = pure st
 
 -- | Data structure to track progress across the project
-data InProgressState = InProgressState
-    { todoVar    :: TVar Int  -- ^ Number of files to do
-    , doneVar    :: TVar Int  -- ^ Number of files done
-    , currentVar :: STM.Map NormalizedFilePath Int
-    }
+data InProgressState
+  = InProgressState
+      { -- | Number of files to do
+        todoVar    :: TVar Int,
+        -- | Number of files done
+        doneVar    :: TVar Int,
+        currentVar :: STM.Map NormalizedFilePath Int
+      }
+  | InProgressStateOutSide
+      { -- | Number of files to do
+        todo :: STM Int,
+        -- | Number of files done
+        done :: STM Int
```

Comment:
This is kind of neat, we just rely on STM transaction retrying to notice when things change. Worth explaining! 

I wonder if we could just use this as the primary mechanism, and then have the other kind of progress reporting as a little wrapper around it?

Reply:
Yes, I've already turned the `lspShakeProgressNew` of `InProgressState` to `LSP.runLspT lspEnv $ withProgress title Nothing NotCancellable $ \update -> loop update 0 (readTVar todoVar) (readTVar doneVar)`
We are using it as the `primary mechanism` now.